### PR TITLE
OHRM5X-277: Admin : Configuration - Register OAuth client

### PR DESCRIPTION
**Pull request by yasirunilan**
This pull request has the changes related to the feature implementation of the Admin Registering OAuth Client. APIs and frontend are both implemented.

### CODE REVIEW

File: symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/EditOAuthClient.vue [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                  :rules="rules.redirectUri"
+                />
+              </oxd-grid-item>
+            </oxd-grid>
+          </oxd-form-row>
+        </div>
+
+        <oxd-form-row>
+          <oxd-grid :cols="1" class="orangehrm-full-width-grid">
+            <oxd-text tag="span" class="orangehrm-link"
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/EditOAuthClient.vue:63]: may have to format by hand.
```
<oxd-text tag="span" class="orangehrm-link">
   API Documentation:
   <a class="orangehrm-link-url"
      href="https://orangehrm.github.io/orangehrm-api-doc"
      >
      https://orangehrm.github.io/orangehrm-api-doc
   </a>
</oxd-text>
```

File: symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/oauth-client.scss [hunk 2]
```diff
@@ -0,0 +1,7 @@
+.orangehrm-link {
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/oauth-client.scss:1]: Please add code header. [refer](https://github.com/orangehrm/orangehrm/blob/5.0/symfony/client/src/orangehrmPimPlugin/pages/employee/employee.scss#L1)

**Review by Super-Chama — APPROVED** [review]
Few suggestions for better code readability.

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            ...$this->getSortingAndPaginationParamsRules(OAuthClientSearchFilterParams::ALLOWED_SORT_FIELDS)
+        );
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function create(): EndpointResult
+    {
+        // TODO:: Check data group permission
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php:101]: if you are keen to add data-group permission .. it's supporting now .. 
anyway .. nice-to-have thing :)

File: symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/EditOAuthClient.vue [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+ -->
+
+<template>
+  <div class="orangehrm-background-container">
+    <div class="orangehrm-card-container">
+      <oxd-text tag="h6" class="orangehrm-main-title"
+        >Edit OAuth Client</oxd-text
```
[the remark this answers is not in the document]
Comment (yasirunilan) [reply symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/EditOAuthClient.vue]: done

File: symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/EditOAuthClient.vue [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+      >
+
+      <oxd-divider />
+
+      <oxd-form novalidate="true" :loading="isLoading" @submitValid="onSave">
+        <div>
```
[the remark this answers is not in the document]
Comment (yasirunilan) [reply symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/EditOAuthClient.vue]: done

File: symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/EditOAuthClient.vue (hunk 1 above)
Comment (yasirunilan) [reply symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/EditOAuthClient.vue]: done

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    /**
+     * @param string $isDefault
+     */
+    public function setIsDefault(string $isDefault): void
+    {
+        $this->isDefault = $isDefault;
+    }
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php:76]: unnecessary blank lines

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+     * @ORM\Column(name="scope", type="string", length=255, nullable=false)
+     */
+    private string $scope;
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(name="is_default", type="boolean", nullable=true, options={"default":0})
+     */
+    private string $isDefault;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php:43]: should `private bool $default`, refactor getters and setters

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+{
+    /**
+     * @var string
+     *
+     * @ORM\Column(name="scope", type="string", length=255, nullable=false)
+     */
+    private string $scope;
+
+    /**
+     * @var string
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php:39]: @var bool

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthClient.php [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * @param string $grantTypes
+     */
+    public function setGrantTypes(string $grantTypes): void
+    {
+        $this->grantTypes = $grantTypes;
+    }
+
+    /**
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthClient.php:133]: move below properties to top of getter/setter method declaration

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthClient.php [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+     * @var string
+     *
+     * @ORM\Column(name="scope", type="string", length=4000)
+     */
+    private string $scope;
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(name="grant_types", type="string", length=80)
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthClient.php:143]: seems `grant_types` and `scope` nullable in db,
```php
@var string|null

private ?string $scope;
```

File: symfony/plugins/orangehrmCoreOAuthPlugin/config/routes.yaml [hunk 11]
```diff
@@ -0,0 +1,33 @@
+apiv2_admin_oauth_client:
+  path: /api/v2/admin/oauth-client
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/config/routes.yaml:2]: can't we change `/api/v2/admin/oauth-client` => `/api/v2/admin/oauth-clients`

File: symfony/plugins/orangehrmCoreOAuthPlugin/Controller/SaveOAuthClientController.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+
+use OrangeHRM\Core\Controller\AbstractVueController;
+use OrangeHRM\Core\Vue\Component;
+use OrangeHRM\Core\Vue\Prop;
+use OrangeHRM\Framework\Http\Request;
+
+class SaveOAuthClientController  extends AbstractVueController
+{
+    public function preRender(Request $request): void
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/Controller/SaveOAuthClientController.php:30]: `@inheritDoc`

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+        );
+    }
+
+    /**
+     * @return OAuthClient
+     * @throws RecordNotFoundException|DaoException
+     */
+    public function saveOAuthClient(): OAuthClient
+    {
+        // TODO:: Check data group permission
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php:200]: will remove this

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+        );
+    }
+
+    /**
+     * @return ParamRule[]
+     */
+    private function getCommonBodyValidationRules(): array
+    {
+        return [
+            new ParamRule(self::PARAMETER_CLIENT_ID),
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php:123]: need to define rules, length validations

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+        return new EndpointResourceResult(ArrayModel::class, $ids);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRuleForDelete(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(CommonParams::PARAMETER_IDS),
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php:145]: better to handle undeletable ids in BE as well.
Can refer 

https://github.com/orangehrm/orangehrm/blob/5.0/symfony/plugins/orangehrmAdminPlugin/Api/UserAPI.php#L246-L263

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php (hunk 3 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php:101]: remove this comment

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getOne(): EndpointResult
+    {
+        $id = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_QUERY, self::PARAMETER_CLIENT_ID);
+        $oAuthClient = $this->getOAuthService()->getOAuthClientByClientId($id);
+        if (!$oAuthClient instanceof OAuthClient) {
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php:156]: replace if with
```php
$this->throwRecordNotFoundExceptionIfNotExist($oAuthClient, OAuthClient::class);
```

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+    public function saveOAuthClient(): OAuthClient
+    {
+        // TODO:: Check data group permission
+        $clientId = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_CLIENT_ID);
+        $existingClientId = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_QUERY, self::PARAMETER_CLIENT_ID);
+        $clientSecret = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_CLIENT_SECRET);
+        $redirectUri = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_REDIRECT_URI);
+        if (!empty($existingClientId)) {
+            $oAuthClient = $this->getOAuthService()->getOAuthClientByClientId($existingClientId);
+            if ($oAuthClient == null) {
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php:207]: replace
```php
$this->throwRecordNotFoundExceptionIfNotExist($oAuthClient, OAuthClient::class);
```

File: symfony/plugins/orangehrmCoreOAuthPlugin/Constant/Scope.php [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * OrangeHRM is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCoreOAuthPlugin/Constant/Scope.php:19]: prefered if we can refactor all files with a blank line after copyright header

**Review by RajithaKumara — CHANGES_REQUESTED** [review]
Great job :tada:  initiating new plugin

File: symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/oauth-client.scss (hunk 2 above)
Comment (yasirunilan) [reply symfony/client/src/orangehrmCoreOAuthPlugin/pages/registerOAuthClient/oauth-client.scss:17]: done

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php (hunk 3 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php]: done

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php (hunk 6 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php]: removed

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php (hunk 7 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php]: updated

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php (hunk 8 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthScope.php]: updated

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthClient.php (hunk 9 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthClient.php]: updated

File: symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthClient.php (hunk 10 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/entity/OAuthClient.php]: updated

File: symfony/plugins/orangehrmCoreOAuthPlugin/Constant/Scope.php (hunk 18 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/Constant/Scope.php:18]: done

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php (hunk 17 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php]: done

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php (hunk 16 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php]: done

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php (hunk 14 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php]: done

File: symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php (hunk 13 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/Api/OAuthClientAPI.php]: done

File: symfony/plugins/orangehrmCoreOAuthPlugin/Controller/SaveOAuthClientController.php (hunk 12 above)
Comment (yasirunilan) [reply symfony/plugins/orangehrmCoreOAuthPlugin/Controller/SaveOAuthClientController.php:33]: done

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**